File: MyProjects/MAP/Assignment 8/src/Model/Statements/CloseRFileStmt.java
package Model.Statements;

import Model.Expressions.IExpr;
import Model.ProgramState;
import Model.Utilities.IDictionary;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.IOException;

public class CloseRFileStmt implements IStmt {
    private IExpr id;
    public CloseRFileStmt(IExpr e){id=e;}

    @Override
    public ProgramState execute(ProgramState state) throws IOException {
        IDictionary<String,Integer> dic=state.getSymTable();
        IDictionary<Integer, BufferedReader> filetable=state.getFileTable();
        IDictionary<Integer, Integer> heap=state.getHeap();
        int val=id.evaluate(dic,heap);
        if(filetable.exists(val)) {
            BufferedReader b = filetable.getValue(val);
            b.close();
            filetable.delete(val);
        }
        else throw new FileNotFoundException("File not found!");
        return null;
    }

    public String toString(){
        return "close("+id+")";
    }
}

File: MyProjects/MAP/Assignment 7/Interpreter/Model/Statements/CloseRFileStmt.cs
using System;
using System.Collections.Generic;
using System.IO;
using Interpreter.Model.Expressions;
using System.Text;

namespace Interpreter.Model.Statements
{
    public class CloseRFileStmt:IStmt
    {
        private IExpr id;

        public CloseRFileStmt(IExpr e) { id = e; }

        public ProgramState execute(ProgramState state)
        {
            int val = id.evaluate(state.SymbolTable);
            if(state.FileTab.exists(val)) {
                StreamReader b = state.FileTab.getValue(val);
            b.Close();
                state.FileTab.delete(val);
            }
            else throw new FileNotFoundException("File not found!");
            return state;
        }

        public override string ToString()
        {
            return "close(" + id + ")";
        }
    }
}
